compat/mingw.c

```diff
@@ -950,20 +950,37 @@ struct tm *localtime_r(const time_t *timep, struct tm *result)
 
 char *mingw_getcwd(char *pointer, int len)
 {
-	int i;
 	wchar_t cwd[MAX_PATH], wpointer[MAX_PATH];
+	DECLARE_PROC_ADDR(kernel32.dll, DWORD, GetFinalPathNameByHandleW,
+			  HANDLE, LPWSTR, DWORD, DWORD);
 	DWORD ret = GetCurrentDirectoryW(ARRAY_SIZE(cwd), cwd);
 
 	if (ret < 0 || ret >= ARRAY_SIZE(cwd))
 		return NULL;
-	ret = GetLongPathNameW(cwd, wpointer, ARRAY_SIZE(wpointer));
-	if (!ret || ret >= ARRAY_SIZE(wpointer))
-		return NULL;
-	if (xwcstoutf(pointer, wpointer, len) < 0)
-		return NULL;
-	for (i = 0; pointer[i]; i++)
-		if (pointer[i] == '\\')
-			pointer[i] = '/';
+
+	if (INIT_PROC_ADDR(GetFinalPathNameByHandleW)) {
+		HANDLE hnd = CreateFileW(cwd, 0,
+			FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
+			OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
```

Comment:
> To open a directory using CreateFile() need to specify the FILE_FLAG_BACKUP_SEMANTICS flag as part of dwFlagsAndAttributes.

Documentation?

> Usually user can open existing directory with such flag.

This is not good enough. You may have talked me into accepting the patch with such a handwaving statement. Before there were so many questions and doubts. But not now.